Use User Locale in Downloads and Exports

adam-james-v opened the pull request:
Fixes #49531

Previously, downloads and exports would always use the site locale setting, now, if the user-locale is set, that is used when formatting the export/download.

adam-james-v committed:
Use User Locale in Downloads and Exports

Fixes #49531

Previously, downloads and exports would always use the site locale setting, now, if the user-locale is set, that is
used when formatting the export/download.

dpsutton commented:
I laid out a different approach earlier. This approach has a fatal flaw I think

Ex:

#### make a request with a korean locale
```shell
❯ http post "$URL?format_rows=true" x-api-key:$API_KEY x-metabase-locale:ko -pb
ID,User ID,Product ID,Subtotal,Tax,Total,Discount ($),Created At,Quantity
1,1,14,37.65,2.07,39.72,,"2월 11, 2025, 9:40 오후",2
2,1,123,110.93,6.1,117.03,,"5월 15, 2024, 8:04 오전",3
3,1,105,52.72,2.9,49.21,6.42,"12월 6, 2025, 10:22 오후",2
```

The response comes back formatted with korean months.

But the instance language looks strange:
<img width="1155" alt="image" src="https://github.com/user-attachments/assets/23053ec2-97cd-43fc-b8b6-c3abb8d19aff">

And that's because you are setting a global locale setting to a user setting:

```clojure
(Locale/setDefault (Locale. (or i18n/*user-locale* (public-settings/site-locale))))
```

We cannot mutate a global setting to whatever _user_ preference might come in.

dpsutton commented:
and easily seen by running the tests:

```clojure
(let [before (.getDisplayLanguage (java.util.Locale/getDefault))]
  (clojure.test/run-test proper-locale-in-exports-test)
  [before (.getDisplayLanguage (java.util.Locale/getDefault))])
[stack trace from some errors locally for me]
Ran 1 tests containing 4 assertions.
1 failures, 2 errors.
["English" "한국어"]
```

adam-james-v committed:
call time formatter with the user locale

Dan correctly points out that we can't set the Default Locale to the user's locale because that sets the whole
instance's default Locale to whatever the last user's locale was. Not good.

Here, I use Dan's original suggestion of using `.withLocale`. In places where t/format is called, we wrap it so that
the user locale is used. In the case where the user locale is not set, the util function defaults to the site-locale
anyway, so things should be ok

adam-james-v committed:
working on getting the test actually working

Set the user's locale and change the util fns to use current user if it's available.

This test has revealed that alerts/subscriptions are not yet properly using the User Locale, so that's something to fix

adam-james-v committed:
still working on this test

adam-james-v committed:
Merge branch 'master' into bugfix-49531-use-user-locale-in-exports

adam-james-v committed:
change up the test to check downloads. subscriptions should use site-locale

adam-james-v committed:
Merge branch 'master' into bugfix-49531-use-user-locale-in-exports

adam-james-v committed:
locale is working for downloads and explain why xlsx is not included in test

adam-james-v committed:
Merge branch 'master' into bugfix-49531-use-user-locale-in-exports
